### PR TITLE
Drop M2Crypto and PyCrypto.

### LINKED ISSUE #164: Cryptography Libraries
As currently structured this project has dependencies on both PyCrypto and M2Crypto. This doesn't seem ideal, especially as the latter is essentially a moribund project with no active path towards py3 compatibility. Is there a plan to move to just one?

Full disclosure: This came to my attention because I work on [pyca/cryptography](https://github.com/pyca/cryptography), which is the underlying dependency for [pyOpenSSL](https://github.com/pyca/pyopenssl) and the [future of PyCrypto](https://lists.dlitz.net/pipermail/pycrypto/2014q3/000814.html). pyca/cryptography itself currently would not fit your needs as our X509 support is incomplete (as is our serialization of asymmetric keys), but we're working on it.

### PR DESCRIPTION
**Pull request by kuba**
Thanks to awesome work of https://github.com/pyca/ folks, I was able to drop `M2Crypto` and `PyCrypto` dependencies in favour of `cryptograpy>=0.8` and `pyopenssl>=0.15`!

Partially fixes #164. Soon to be released `cryptography>=1.0` is blocking further changes to our code (e.g. `x509.CSR.public_bytes()` or `x509.Cert.public_bytes()` necessary in `acme`).

This PR builds on top of #586 and due to very likely conflicts blocks further dev process. Could you please review, @jdkasten?

cc: @reaperhulk @alex https://github.com/pyca/cryptography/issues/1957

### CODE REVIEW

File: letsencrypt/proof_of_possession.py [hunk 1]
```diff
@@ -39,19 +45,28 @@ def perform(self, achall):
...
+            cert_key = achall.alg.kty(key=jose.ComparableRSAKey(
+                cert_obj.public_key()))
             if cert_key == achall.hints.jwk:
                 return self._gen_response(achall, key)
 
         # Is there are different prompt we should give the user?
         code, key = zope.component.getUtility(
             interfaces.IDisplay).input(
-                "Path to private key for identifier: %s " % achall.domain)
+                "QPath to private key for identifier: %s " % achall.domain)
```
Comment (kuba) [inline letsencrypt/proof_of_possession.py]: typo

File: acme/jose/jwa.py [hunk 2]
```diff
@@ -66,43 +73,79 @@ class _JWAHS(JWASignature):
...
-
-        .. warning::
-            Does not protect against timing attack (no constant compare).
-
-        """
-        return self.sign(key, msg) == sig
+        verifier = hmac.HMAC(key, self.hash, backend=default_backend())
+        verifier.update(msg)
+        try:
+            verifier.verify(sig)
```
Comment (jdkasten) [inline acme/jose/jwa.py]: Great!

File: acme/jose/jwk.py [hunk 3]
```diff
@@ -120,21 +122,73 @@ def load(cls, string):
         :rtype: :class:`JWKRSA`
```
Comment (jdkasten) [inline acme/jose/jwk.py]: Probably worth specifying that the RSA key string should be in PEM form.

File: acme/jose/jwk.py [hunk 4]
```diff
@@ -120,21 +122,73 @@ def load(cls, string):
...
 
         """
-        return cls(key=util.HashableRSAKey(
-            Crypto.PublicKey.RSA.importKey(string)))
+        try:
+            key = serialization.load_pem_public_key(
```
Comment (jdkasten) [inline acme/jose/jwk.py]: I haven't gotten a chance to test this, but does the private key not provide all of the functions of the private key?
My inclination would be to try to load it as a private first and then as a public.  I will try to read up on exactly what is going on here.

File: bootstrap/_deb_common.sh [hunk 5]
```diff
@@ -44,20 +44,14 @@ else
...
-#           #276, https://github.com/martinpaljak/M2Crypto/issues/62,
-#           M2Crypto setup.py:add_multiarch_paths
-
 apt-get install -y --no-install-recommends \
   git-core \
   python \
   python-dev \
   "$virtualenv" \
   gcc \
-  swig \
```
Comment (jdkasten) [inline bootstrap/_deb_common.sh]: :+1:

File: letsencrypt/client.py [hunk 6]
```diff
@@ -431,8 +434,10 @@ def validate_key_csr(privkey, csr=None):
 
     if csr:
         if csr.form == "der":
-            csr_obj = M2Crypto.X509.load_request_der_string(csr.data)
-            csr = le_util.CSR(csr.file, csr_obj.as_pem(), "der")
```
Comment (jdkasten) [inline letsencrypt/client.py]: This looks like you removed an error.  Nice catch

File: letsencrypt/crypto_util.py [hunk 7]
```diff
@@ -309,3 +326,19 @@ def get_sans_from_csr(csr, typ=OpenSSL.crypto.FILETYPE_PEM):
...
+
+
+def asn1_generalizedtime_to_dt(timestamp):
+    """Convert ASN.1 GENERALIZEDTIME to datetime.
+
+    Useful for deserialization of `OpenSSL.crypto.X509.get_notAfter` and
+    `OpenSSL.crypto.X509.get_notAfter` outputs.
+
+    """
+    # TODO: there are two other ts formats
```
Comment (jdkasten) [inline letsencrypt/crypto_util.py]: This TODO should move up into the function docstring.  It seems to me to be pretty important.

File: letsencrypt/plugins/standalone/tests/authenticator_test.py [hunk 8]
```diff
@@ -400,7 +394,7 @@ def test_start_listener_fork_child(self, mock_fork, mock_atfork):
         self.assertEqual(self.authenticator.child_pid, os.getpid())
         self.authenticator.do_child_process.assert_called_once_with(
             1717, "key")
-        mock_atfork.assert_called_once_with()
```
Comment (jdkasten) [inline letsencrypt/plugins/standalone/tests/authenticator_test.py]: Reminder: Take a closer look to make sure tests are still functioning correctly

File: letsencrypt/revoker.py [hunk 9]
```diff
@@ -13,11 +13,11 @@
 import shutil
 import tempfile
 
-import Crypto.PublicKey.RSA
-import M2Crypto
+import OpenSSL
```
Comment (jdkasten) [inline letsencrypt/revoker.py]: Thanks for updating... sorry for the dead code :(

File: letsencrypt/crypto_util.py (hunk 7 above)
Comment (kuba) [inline letsencrypt/crypto_util.py]: Done

File: acme/jose/jwk.py (hunk 4 above)
Comment (kuba) [inline acme/jose/jwk.py]: I moved this code to `JWK.load` and restructured a bit; Privkeys in PEM/DER are tried first, then pubkeys PEM/DER.

File: acme/jose/jwk.py (hunk 3 above)
Comment (kuba) [inline acme/jose/jwk.py]: This is fixed together with move to `JWK.load`

**Comment by kuba** [comment]
Will reopen.

**Pull request closed without merging**